### ISSUE TITLE
Assertion 'Symbol' failed with -fpatchable-function-entry=2 + -Oz AArch64

### ISSUE THREAD

**Issue by nathanchance** [llvm/llvm-project#52635]
Initial report: https://github.com/ClangBuiltLinux/linux/issues/1523

When building the Linux kernel with `-Oz`, there is a failed assertion: ```llvm/lib/MC/MCExpr.cpp:207: llvm::MCSymbolRefExpr::MCSymbolRefExpr(const llvm::MCSymbol *, llvm::MCSymbolRefExpr::VariantKind, const llvm::MCAsmInfo *, llvm::SMLoc): Assertion `Symbol' failed.```

C reproducer:

```c
enum { false } cs47l85_16bit_readable_register() { return false; }
_Bool cs47l85_16bit_volatile_register() { return false; }
```

```
$ clang --target=aarch64-linux-gnu -fpatchable-function-entry=2 -c -o /dev/null test.i

$ clang --target=aarch64-linux-gnu -Oz -c -o /dev/null test.i

$ clang --target=aarch64-linux-gnu -fpatchable-function-entry=2 -Oz -c -o /dev/null test.i
clang: /home/nathan/cbl/tc-build/llvm-project/llvm/lib/MC/MCExpr.cpp:207: llvm::MCSymbolRefExpr::MCSymbolRefExpr(const llvm::MCSymbol *, llvm::MCSymbolRefExpr::VariantKind, const llvm::MCAsmInfo *, llvm::SMLoc): Assertion `Symbol' failed.
...
```

LLVM IR reproducer:

```
target triple = "aarch64-unknown-linux-gnu"

; Function Attrs: minsize
define i32 @cs47l85_16bit_readable_register() #0 {
entry:
  ret i32 0
}

; Function Attrs: minsize
define i1 @cs47l85_16bit_volatile_register() #0 {
entry:
  ret i1 false
}

attributes #0 = { minsize "patchable-function-entry"="2" }
```

```
$ llc reduced.ll
llc: /home/nathan/cbl/tc-build/llvm-project/llvm/lib/MC/MCExpr.cpp:207: llvm::MCSymbolRefExpr::MCSymbolRefExpr(const llvm::MCSymbol *, llvm::MCSymbolRefExpr::VariantKind, const llvm::MCAsmInfo *, llvm::SMLoc): Assertion `Symbol' failed.
PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace.
Stack dump:
0.      Program arguments: llc reduced.ll
1.      Running pass 'Function Pass Manager' on module 'reduced.ll'.
2.      Running pass 'AArch64 Assembly Printer' on function '@OUTLINED_FUNCTION_0'
 #0 0x0000000004149104 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/usr/local/stow/llvm/2021-12-02_15-48-31-e45705ad5051247e1cb2355094ef77316ae24c37/bin/llc+0x4149104)
 #1 0x000000000348dd4b (/usr/local/stow/llvm/2021-12-02_15-48-31-e45705ad5051247e1cb2355094ef77316ae24c37/bin/llc+0x348dd4b)
 #2 0x00007f1d38806870 __restore_rt sigaction.c:0:0
 #3 0x00007f1d38220d22 raise (/usr/lib/libc.so.6+0x3cd22)
 #4 0x00007f1d3820a862 abort (/usr/lib/libc.so.6+0x26862)
 #5 0x00007f1d3820a747 _nl_load_domain.cold loadmsgcat.c:0:0
 #6 0x00007f1d38219616 (/usr/lib/libc.so.6+0x35616)
 #7 0x000000000263e4d6 llvm::MCSymbolRefExpr::create(llvm::MCSymbol const*, llvm::MCSymbolRefExpr::VariantKind, llvm::MCContext&, llvm::SMLoc) (/usr/local/stow/llvm/2021-12-02_15-48-31-e45705ad5051247e1cb2355094ef77316ae24c37/bin/llc+0x263e4d6)
 #8 0x00000000029d5034 llvm::AsmPrinter::emitXRayTable() (/usr/local/stow/llvm/2021-12-02_15-48-31-e45705ad5051247e1cb2355094ef77316ae24c37/bin/llc+0x29d5034)
 #9 0x00000000029d4875 (/usr/local/stow/llvm/2021-12-02_15-48-31-e45705ad5051247e1cb2355094ef77316ae24c37/bin/llc+0x29d4875)
#10 0x00000000029894c3 llvm::MachineFunctionPass::runOnFunction(llvm::Function&) (/usr/local/stow/llvm/2021-12-02_15-48-31-e45705ad5051247e1cb2355094ef77316ae24c37/bin/llc+0x29894c3)
#11 0x0000000002504e8b llvm::FPPassManager::runOnFunction(llvm::Function&) (/usr/local/stow/llvm/2021-12-02_15-48-31-e45705ad5051247e1cb2355094ef77316ae24c37/bin/llc+0x2504e8b)
#12 0x000000000250489d llvm::FPPassManager::runOnModule(llvm::Module&) (/usr/local/stow/llvm/2021-12-02_15-48-31-e45705ad5051247e1cb2355094ef77316ae24c37/bin/llc+0x250489d)
#13 0x0000000002503e77 llvm::legacy::PassManagerImpl::run(llvm::Module&) (/usr/local/stow/llvm/2021-12-02_15-48-31-e45705ad5051247e1cb2355094ef77316ae24c37/bin/llc+0x2503e77)
#14 0x000000000370c169 main (/usr/local/stow/llvm/2021-12-02_15-48-31-e45705ad5051247e1cb2355094ef77316ae24c37/bin/llc+0x370c169)
#15 0x00007f1d3820bb25 __libc_start_main (/usr/lib/libc.so.6+0x27b25)
#16 0x000000000370993e _start (/usr/local/stow/llvm/2021-12-02_15-48-31-e45705ad5051247e1cb2355094ef77316ae24c37/bin/llc+0x370993e)
```

My bisect pointed to https://github.com/llvm/llvm-project/commit/10bc12588dac532fad044b2851dde8e7b9121e88, cc @MaskRay

**Comment by MaskRay** [comment]
This is just a lurking problem exposed by the XRay change. It would have weird effects without the XRay change.

MachineOutliner creates `OUTLINED_FUNCTION_0` which does not have the `patchable-function-entry` but whose body has a `TargetOpcode::PATCHABLE_FUNCTION_ENTER` MachineInstr.

`patchable-function-entry` can mean either XRay instrumentation or `-fpatchable-function-entry=`. In the absence of the `patchable-function-entry` function attribute, it is interpreted as XRay instrumentation. Since `patchable-function-entry` is absent, `CurrentFnBegin` is nullptr and we have an null assert failure.

**Comment by nathanchance** [comment]
Reopening to track adding https://github.com/llvm/llvm-project/commit/a6a07a514b8a084feaa7f4f15d569698b9840d83 to `release/13.x`.

**Comment by tstellar** [comment]
This fix does not apply cleanly to the release branch.  Could someone resolve the merge conflicts and push it to a github fork.

**Comment by nathanchance** [comment]
@tstellar https://github.com/nathanchance/llvm-project/commit/2ffa7a3bc7e14ad62a2636a07b64a5bd85433cc7

I am running it through my tests now.

**Comment by nathanchance** [comment]
That change builds, passed `check-clang`, and the original test case.

**Comment by tstellar** [comment]
@MaskRay Is this safe to backport?

**Comment by MaskRay** [comment]
It is safe. Please backport :)

**Comment by tstellar** [comment]
Merged: 0d44201451f0
